File: kotlin/src/main/kotlin/com/authdog/sdk/AuthdogClient.kt
package com.authdog.sdk

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.decodeFromString
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.net.URI
import java.time.Duration
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

/**
 * Configuration for the Authdog client
 */
data class AuthdogClientConfig(
    val baseUrl: String,
    val apiKey: String? = null,
    val timeoutSeconds: Long = 10,
    val httpClient: HttpClient? = null
)

/**
 * Main client for interacting with Authdog API
 */
class AuthdogClient(private val config: AuthdogClientConfig) {
    private val httpClient: HttpClient = config.httpClient ?: HttpClient.newBuilder()
        .connectTimeout(Duration.ofSeconds(config.timeoutSeconds))
        .build()
    
    private val json = Json {
        ignoreUnknownKeys = true
        coerceInputValues = true
    }

    /**
     * Get user information using an access token
     * @param accessToken The access token for authentication
     * @return UserInfoResponse containing user information
     * @throws AuthenticationError When authentication fails
     * @throws APIError When API request fails
     */
    suspend fun getUserInfo(accessToken: String): UserInfoResponse = withContext(Dispatchers.IO) {
        val url = "${config.baseUrl.removeSuffix("/")}/v1/userinfo"
        
        val request = HttpRequest.newBuilder()
            .uri(URI.create(url))
            .header("Content-Type", "application/json")
            .header("User-Agent", "authdog-kotlin-sdk/0.1.0")
            .header("Authorization", "Bearer $accessToken")
            .GET()
            .build()

        val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString())
        
        when (response.statusCode()) {
            200 -> {
                try {
                    json.decodeFromString<UserInfoResponse>(response.body())
                } catch (e: Exception) {
                    throw APIError("Failed to parse response: ${e.message}")
                }
            }
            401 -> throw AuthenticationError("Unauthorized - invalid or expired token")
            500 -> {
                try {
                    val errorResponse = json.decodeFromString<ErrorResponse>(response.body())
                    when (errorResponse.error) {
                        "GraphQL query failed" -> throw APIError("GraphQL query failed")
                        "Failed to fetch user info" -> throw APIError("Failed to fetch user info")
                        else -> throw APIError("HTTP error 500: ${response.body()}")
                    }
                } catch (e: Exception) {
                    throw APIError("HTTP error 500: ${response.body()}")
                }
            }
            else -> throw APIError("HTTP error ${response.statusCode()}: ${response.body()}")
        }
    }

    /**
     * Close the HTTP client (for cleanup)
     */
    fun close() {
        // HttpClient doesn't require explicit cleanup in Java/Kotlin
    }
}

/**
 * Error response from the API
 */
@Serializable
data class ErrorResponse(
    val error: String
)

File: scala/src/main/scala/com/authdog/AuthdogClient.scala
package com.authdog

import com.authdog.exceptions.{ApiException, AuthenticationException}
import com.authdog.types.UserInfoResponse
import sttp.client3._
import sttp.client3.circe._
import io.circe.generic.auto._
import scala.concurrent.duration._
import scala.util.Try

/**
 * Main client for interacting with Authdog API
 */
class AuthdogClient(
  baseUrl: String,
  apiKey: Option[String] = None,
  timeout: Duration = 10.seconds
) {
  private val backend = HttpURLConnectionBackend()
  private val cleanBaseUrl = baseUrl.stripSuffix("/")
  
  /**
   * Get user information using an access token
   * @param accessToken The access token for authentication
   * @return UserInfoResponse containing user information
   * @throws AuthenticationException When authentication fails
   * @throws ApiException When API request fails
   */
  def getUserInfo(accessToken: String): UserInfoResponse = {
    val request = basicRequest
      .get(uri"$cleanBaseUrl/v1/userinfo")
      .header("Content-Type", "application/json")
      .header("User-Agent", "authdog-scala-sdk/0.1.0")
      .header("Authorization", s"Bearer $accessToken")
      .readTimeout(timeout)
      .response(asJson[UserInfoResponse])
    
    // Add API key if provided
    val finalRequest = apiKey match {
      case Some(key) => request.header("Authorization", s"Bearer $key")
      case None => request
    }
    
    val response = finalRequest.send(backend)
    
    response.body match {
      case Right(userInfo) => userInfo
      case Left(error) =>
        response.code.code match {
          case 401 => throw AuthenticationException("Unauthorized - invalid or expired token")
          case 500 => 
            // Try to parse error response
            Try {
              import io.circe.parser._
              parse(error.toString).flatMap(_.hcursor.get[String]("error")).toOption match {
                case Some("GraphQL query failed") => throw ApiException("GraphQL query failed")
                case Some("Failed to fetch user info") => throw ApiException("Failed to fetch user info")
                case _ => throw ApiException(s"HTTP error 500: $error")
              }
            }.getOrElse(throw ApiException(s"HTTP error 500: $error"))
          case statusCode => throw ApiException(s"HTTP error $statusCode: $error")
        }
    }
  }
  
  /**
   * Close the HTTP client (for cleanup)
   */
  def close(): Unit = {
    backend.close()
  }
}

object AuthdogClient {
  /**
   * Create a new Authdog client
   * @param baseUrl The base URL of the Authdog API
   * @return AuthdogClient instance
   */
  def apply(baseUrl: String): AuthdogClient = new AuthdogClient(baseUrl)
  
  /**
   * Create a new Authdog client with API key
   * @param baseUrl The base URL of the Authdog API
   * @param apiKey Optional API key for authentication
   * @return AuthdogClient instance
   */
  def apply(baseUrl: String, apiKey: String): AuthdogClient = new AuthdogClient(baseUrl, Some(apiKey))
  
  /**
   * Create a new Authdog client with custom timeout
   * @param baseUrl The base URL of the Authdog API
   * @param apiKey Optional API key for authentication
   * @param timeout Request timeout
   * @return AuthdogClient instance
   */
  def apply(baseUrl: String, apiKey: Option[String], timeout: Duration): AuthdogClient = 
    new AuthdogClient(baseUrl, apiKey, timeout)
}
